feature_details.md

```diff
@@ -0,0 +1,413 @@
+
+An important assumption in the design of this feature is that it's reasonable to limit customization options for cross-origin transitions to make it easier for UAs to enforce the constraints above. But the API/implementation should strive to not unnecessarily limit capabilities for same-origin transitions due to cross-origin constraints.
+
+## Glossary
+| Term | Description |
+| ------------- | ------------- |
+| Previous Document  | The Document the user is viewing when a navigation is initated.  |
+| New Document  | The Document which will be current in the session history when a navigation is committed. |
+| Root transition | An animation to transition the content for root element of both Documents. |
+| Exit transition | An animation which specifies how an element in the previous Document exits the screen. |
+| Enter transition | An animation which specifies how an element in the new Document enters the screen. |
+| Snapshot | A pixel copy for a DOM subtree. |
+
+## Sample Code
+This section explains how the feature can be used to add transitions to MPAs. There are a set of use-cases and edge cases to consider :
+
+### Root Transition
+Let's start with a root transition, which adds an animation to the root element of previous and new Documents. Below is a video of the transition we want to create :
+
+[![Video Link for Root Element Transition](https://img.youtube.com/vi/mCvvWMIGItY/0.jpg)](https://www.youtube.com/watch?v=mCvvWMIGItY)
+
+The entry point for this API is a new `document.documentTransition` object. The previous Document provides a list of elements which will be animated during the transition. Since this is root transition, this list only includes the `html` element. The result of calling `documentTransition.setSameOriginTransitionElements` as illustrated below is that the browser caches a copy of the pixels (snapshot) for the `html` element when a navigation to `foo.com/a.html` is initiated.
+
+```
+// The list of elements is keyed on url. When a user navigates to
+// a url, the corresponding entry in this dictionary is used to
+// decide which elements should be cached.
+document.documentTransition.setSameOriginTransitionElements(
+  {
+   “foo.com/a.html” : [{element : document.html}],
+  });
+```
+
+Now we need to add some code on the new Document. When the new Document loads, script registers a listener for a new event type called `handleTransition`. The event is dispatched if a transition was initiated by the previous Document. The exact Document lifecycle stage before which this event should be registered is TBD.
+
+```
+document.documentTransition.addEventListener(
+    “handleTransition”, e => {
+        e.startTransition(provideTransitionFrom(e.previousUrl, e.previousElements));
+    });
+
+async function provideTransitionFrom(previousUrl, previousElements) {
+    // Wait for the image on the new page to load before
+    // starting the transition.
+    await waitForImageLoad();
```

Comment:
In order for this code snippet to do what I expect it wants to do, the page has to still be displaying the previous page's content while this line (and the image itself) is loading. I think it may be possible to delay presenting new content in the same way Chromium does with paint holding, but I'm just curious if we intend for this API to block presentation of new content even further (and for arbitrarily long).

Reply:
The intent indeed is to use paint holding to allow the API to be used to block presentation of new content. I think it's reasonable for the browser to have a timeout (currently we use 4 seconds for the SPA) and how long that should be is up for discussion. But my thinking is that the way an SPA can completely control the experience of delaying transitions between documents, an MPA should be able to do it as well.